README
Perl Module Lingua::Translit
----------------------------

ABOUT

Lingua::Translit can be used to convert text from one writing system to
another, based on national or international transliteration tables.
Where possible a reverse transliteration is supported.


PREREQUISTES

Perl 5.8.0 or above

If you are going to work on the included XML definitions or even want to
write your own, XML::LibXML is needed as well.


INSTALLATION

To test and install Lingua::Translit, untar the archive and execute

	perl Makefile.PL
	make
	make test
	make install


CONTACT

You are welcome to contact the authors at perl@lingua-systems.com
if you have any suggestions and contributions.


COPYRIGHT AND LICENSE

Copyright (C) 2007-2008 Alex Linke and Rona Linke
Copyright (C) 2009-2014 Lingua-Systems Software GmbH

This module is free software. It may be used, redistributed
and/or modified under the terms of either the GPL v2 or the
Artistic license.


Changes
0.21  -- 2014-10-24

    * Added "Common ARA" Arabic transliteration. Thanks to Ahmed Elsheshtawy
      for suggesting this transliteration and his help implementing it!

0.20  -- 2014-05-22

    * lib/Lingua/Translit/Tables.pm: Added sub to handle Perl's
      "Unicode Bug", see perlunicode for details (v0.10). This fixes the
      errors reported by Perl v5.18.0.
    * MANIFEST: Added missing "ALA-LC RUS" files to distribution.
    * xml/: Removed "Common Classical MON" transliteration due to errors
      and contributors lack of time to fix these.
    * Code cleanups

0.19  -- 2011-04-15

    * Added "ALA-LC RUS" transliteration (ALA-LC:1997). Thanks to Dmitry Smal
      and Rusar Publishing for contributing this table!

0.18  -- 2010-07-29

    * Added transliterations:
        * "GOST 7.79 RUS OLD" (GOST 7.79:2000 B, support for Old Russian)
        * "GOST 7.79 RUS"     (GOST 7.79:2000 B, reversible)
        * "GOST 7.79 UKR"     (GOST 7.79:2000 B, reversible)
    * lib/Lingua/Translit/Tables.pm: Improved translit_list_supported()
      output (v0.9)
    * translit: Minor improvements (v0.4)
    * tools/substitute_tables.pl: Typo fixed

0.17  -- 2010-01-20

    * Added "DIN 1460 UKR" transliteration
    * xml/din_1460_rus.xml: Hard and soft sign improvements
    * t/23_tr_DIN_1460_RUS.t: Added all caps hard sign test

0.16  -- 2010-01-04

    * Added "DIN 1460 RUS" transliteration
    * MANIFEST.SKIP: Added to ignore .git
    * t/01_Lingua-Translit-Tables.t: Added various tests to verify any
      supported transliteration can be referenced by a couple of variants
      of their names, i.e. "Common DEU", "common_deu", "COMMON DEU" or
      "common deu".
    * t/11_tr_Common_DEU.t: Added decoded UTF-8 input tests
    * Replaced HTML developer manual with a PDF version

0.15  -- 2009-10-19

    * Added support for the current development version of Perl (5.11) by
      using named Unicode properties instead of POSIX character classes
      within any context dependent transliteration rule.
      The following transliteration tables are affected:
        * xml/din_1460_bul.xml
        * xml/iso_9.xml
        * xml/streamlined_system_bul.xml
      Closes #50613. Thanks to Andreas Koenig for reporting this issue.
    * xml/template.xml: Added some more examples, added missing closing tag.
    * lib/Lingua/Translit/Tables.pm, xml/xml2dump.pl: To increase robustness
      and ease of use, tables are now associated with an identifier instead
      of their names.
    * Removed "utf8" and "bytes" pragmas.
    * Updated indent of all transliteration tables.

0.14  -- 2009-10-16

    * Added Win32 support. (Closes #49544)
      Please note that a Unix-like environment is still required to develop
      transliteration tables using the provided set of tools.
    * t/03_translit.t: Use $^X rather than relying on the shebang
    * Various trivial fixes, code cleanups and hyperlink updates
    * xml/manual/manual.pod: Added missing markup

0.13  -- 2008-04-29

    * Added developer manual covering the addition of own transliteration
      tables (xml/manual/).
    * Added test template (t/xx_tr_template.t.pl)
    * Makefile.PL:
        * Added "manual" target
        * Suited MSWin32 check for CPAN Testers tools

0.12 -- 2008-04-23

    * Added common Polish transliteration ("Common POL")

0.11 -- 2008-04-21

    * xml/din_1460_bul.xml: changed transliteration of SOFT SIGN to plain
      ASCII apostrophe.
    * New transliterations:
        * Common SLK (Slovak)
        * Common SLV (Slovenian)

0.10 -- 2008-04-17

    * lib/Lingua/Translit.pm, xml/translit.dtd: Contexts can now contain two
      patterns at a time (before and after). These are combined using AND
      logics and therefore act as a logical "inbetween".
    * xml/xml2dump.pl: Switched from XML::Simple to XML::LibXML in order to
      prevent loss of whitespace charactes like U+180E.
    * xml/greeklish.xml: GREEK CAPITAL LETTER UPSILON WITH DIALYTIKA (U+03AB)
      bug fixed.
    * t/13_tr_Greeklish.t: Added test for upsilon bugfix.
    * Added two transliterations for Bulgarian:
        * DIN 1460
        * Streamlined System

0.09 -- 2008-04-09

    * Added Mongolian transliteration "Common Classical MON". Thanks to
      Bayanzul Lodoysamba (baynaa@users.sourceforge.net) for contributing
      this table!
    * lib/Lingua/Translit.pm: "from" and "to" transliteration patterns are
      now evaluated literally only (within \Q and \E).
    * lib/Lingua/Translit/Tables.pm: Removed verbose package information from
      translit_list_supported()
    * translit: Display Lingua::Translit's version if show_list() is called
    * translit: For convenience, every underscore in a transliteration's name
      is substituted by a blank.
    * Makefile.PL: Exit with a hint on MSWin32 rather than dying in
      t/03_translit.t
    * xml/xml2dump.pl: workaround for discarded U+180E character (see l.85)
    * Added a transliteration table template (xml/template.xml)

0.08 -- 2008-03-31

    * Added common Czech transliteration "Common CES"

0.07 -- 2008-03-21

    * Fixed a wrong codepoint in an ISO 843 rule.

0.06 -- 2008-03-21

    * Fixed a wrong codepoint in a DIN 31634 rule.

0.05 -- 2008-03-10

    * Renamed "DIN 5008" to "Common DEU" as the DIN norm doesn't cover the
      transliteration explicitly.

0.04 -- 2008-02-28

    * Added common Romanian transliteration ("Common RON")
    * Updated Makefile.PL to use tools/substitute_tables.pl to read a
      transliteration table dump and substitute a placeholder line in
      lib/Lingua/Translit/Tables.pm with it.

0.03 -- 2008-02-25

    * Added DIN 31634 (1982)
    * Test fixes (require Perl 5.8 if the "utf8" pragma is used)

0.02 -- 2008-02-18

    * Added Greeklish
    * Code cleanup

0.01 -- 2008-02-13

    * Initial release
    * Set of supported transliterations:
        * ISO 9
        * ISO 843
        * DIN 5008

# vim: set sw=4 sts=4 ts=4 ai et ft=changelog:


t/xx_tr_template.t.pl
use strict;
use Test::More tests => 3;  # number of tests

my $name	=   "";	    # transliterations name
my $reversible	=   0;	    # is the transliteration reversible?

my $input	=   "";	    # short corpus...
my $output_ok	=   "";	    # ...its correct transliteration

my $context	=   "";	    # context-sensitive example
my $context_ok	=   "";	    # ...its correct transliteration

use Lingua::Translit;

my $tr = new Lingua::Translit($name);


my $output = $tr->translit($input);

# 1
is($tr->can_reverse(), $reversible, "$name: reversibility");

# 2
is($output, $output_ok, "$name: transliteration");

$output = $tr->translit($context);

# 3
is($output, $context_ok, "$name: transliteration (context-sensitive)");

# vim: sts=4 enc=utf-8


xml/xml2dump.pl
#!/usr/bin/perl -w

#
# Copyright (C) 2007-2008 Alex Linke <alinke@lingua-systems.com>
# Copyright (C) 2009-2014 Lingua-Systems Software GmbH
#

use strict;
use warnings;

require 5.008;

use XML::LibXML;
use Data::Dumper;
use Getopt::Long;

my $VERSION = '0.5';

my %tables;

# set default options
my %opt = (
    output  => "tables.dump",
    verbose => 0,
);

# parse commandline options
show_help(1)
  unless GetOptions(
    "output|o=s" => \$opt{output},
    "verbose|v"  => \$opt{verbose},
    "help|h"     => \$opt{help}
  );
show_help(1) if scalar(@ARGV) == 0;    # No XML file(s) given
show_help(0) if $opt{help};

my $xmlparser = new XML::LibXML();

# Set parser options
$xmlparser->pedantic_parser(1);
$xmlparser->validation(1);
$xmlparser->expand_entities(1);
$xmlparser->keep_blanks(1);
$xmlparser->line_numbers(1);

# Treat everything else in @ARGV as a filename
foreach my $file (@ARGV) {
    print "Parsing $file..." if $opt{verbose};

    my %counts = ( rules => 0, contexts => 0 );

    my $ds;

    my $doc = $xmlparser->parse_file($file)
      or die "Error parsing $file: $!\n";

    # Retrieve meta-documentation from XML document first
    foreach my $meta (qw/name desc reverse/) {
        my @nodes = $doc->findnodes("/translit/$meta");

        die "#/translit/$meta != 1" if ( scalar(@nodes) != 1 );

        $ds->{$meta} = $nodes[0]->to_literal();
    }

    # Perform some basic meta data checks
    die "Name undefined.\n"          unless $ds->{name};
    die "Description undefined.\n"   unless $ds->{desc};
    die "Reversibility undefined.\n" unless $ds->{reverse};

    # Check <reverse> tag contains valid data.
    # TODO: move this to the DTD
    die "Reversibility: '$ds->{reverse}' -- Should be 'true' or 'false'.\n"
      unless $ds->{reverse} =~ /^(true|false)$/;

    # Set the table's identifier
    $ds->{id} = lc( $ds->{name} );
    $ds->{id} =~ s/\s/_/g;

    # Retrieve all rules, extract their data and store it to an appropriate
    # data structure
    foreach my $rule ( $doc->findnodes("/translit/rules/rule") ) {
        my @nodes;
        my $rule_ds;

        # Retrieve "from" and "to" literals
        foreach my $n (qw/from to/) {
            @nodes = $rule->findnodes("./$n");

            die "#/translit/rules/rules/$n != 1 "
              . "(at line "
              . $rule->line_number() . ")\n"
              if ( scalar(@nodes) != 1 );

            $rule_ds->{$n} = $nodes[0]->to_literal();
        }

        # Retrieve rule's "context"
        @nodes = $rule->findnodes("./context");

        die "#/translit/rules/rule/context > 1 "
          . "(at line "
          . $rule->line_number() . ")\n"
          if ( scalar(@nodes) > 1 );

        # Process rule's "context" if necessary
        if ( scalar(@nodes) ) {
            foreach my $context (qw/before after/) {
                @nodes = $rule->findnodes("./context/$context");

                die "#/translit/rules/rule/context/$context > 1 "
                  . "(at line "
                  . $rule->line_number() . ")\n"
                  if ( scalar(@nodes) > 1 );

                # Copy the context to the rule's data structure
                if ( scalar(@nodes) ) {
                    $rule_ds->{context}->{$context} = $nodes[0]->to_literal();
                }
            }

            $counts{contexts}++;
        }

        $counts{rules}++;

        die $rule_ds->{name} . ": from==to -> " . $rule_ds->{from} . "\n"
          if ( $rule_ds->{from} eq $rule_ds->{to} );

        push @{ $ds->{rules} }, $rule_ds;
    }

    # Copy transliteration structure over to the final hash
    $tables{ $ds->{id} } = $ds;

    print " ($ds->{id}: rules=$counts{rules}, contexts=$counts{contexts})\n"
      if $opt{verbose};

    undef($ds);    # free memory
}

# Configure Data::Dumper
my $dumper = new Data::Dumper( [ \%tables ], [qw/*tables/] );
$dumper->Purity(0);
$dumper->Useqq(1);
$dumper->Indent(1);

# Dump the table(s) to disk
open FH, ">$opt{output}" or die "$opt{output}: $!\n";
print FH $dumper->Dump();
close(FH);

print scalar( keys(%tables) ),
  " transliteration table(s) dumped to $opt{output}.\n"
  if $opt{verbose};

sub show_help {
    my $retval = shift();

    print STDERR
      "xml2dump v$VERSION -- Copyright 2007-2008 by Alex Linke ",
      "<alinke\@lingua-systems.com>\n\n",
      "usage: $0  [-v -h]  -o FILE  XML-FILE(s)\n\n",
      "\t--output  -o  FILE     set output file (default: transtbl.dump)\n",
      "\t--verbose -v           be verbose\n",
      "\t--help    -h           show this help\n";

    exit($retval);
}

# vim: sw=4 sts=4 ts=4 ai et


tools/substitute_tables.pl
#!/usr/bin/perl -w

#
# Copyright (C) 2007-2008 Alex Linke <alinke@lingua-systems.com>
# Copyright (C) 2009-2014 Lingua-Systems Software GmbH
#

use strict;
use IO::File;

my $tbl_file = 'xml/tables.dump';
my $infile = $ARGV[0] || die "usage: $0 file";

my $fh = new IO::File();

local $/;

# read input file
$fh->open($infile) or die "$infile: $!\n";
my $in_content = <$fh>;
$fh->close();

# read tables file
$fh->open($tbl_file) or die "$tbl_file: $!\n";
my $tbls = <$fh>;
$fh->close();

if ( $in_content =~ s/\n\%tables;\s+# PLACEHOLDER\s*\n/\n$tbls\n/ ) {
    print "$infile: substituted tables: " . length($tbls) . " bytes.\n";
}
else {
    print "$infile: no substitution.\n";
    exit 1;
}

chmod 0644, $infile or die "chmod: $!\n";

# write output to input file
$fh->open("> $infile") or die "$infile: $!\n";
print $fh $in_content;
$fh->close();

# vim: sts=4 sw=4 ts=4 ai et


lib/Lingua/Translit/Tables.pm
package Lingua::Translit::Tables;

#
# Copyright (C) 2007-2008 ...
#   Alex Linke <alinke@lingua-systems.com>
#   Rona Linke <rlinke@lingua-systems.com>
# Copyright (C) 2009-2014 Lingua-Systems Software GmbH
#

use strict;
use warnings;
use utf8;

require 5.008;

our $VERSION = '0.10';

use Carp;

=pod

=head1 NAME

Lingua::Translit::Tables - provides transliteration tables

=head1 SYNOPSIS

  use Lingua::Translit::Tables qw/:checks/;

  my $truth;

  $truth = translit_supported("ISO 9");
  $truth = translit_reverse_supported("ISO 9");

  use Lingua::Translit::Tables qw/:list/;

  translit_list_supported();

=head1 DESCRIPTION

This module is primary used to provide transliteration tables for
L<Lingua::Translit> and therefore allows to separate data and algorithm.

Beyond that, it provides routines to check if a given transliteration is
supported and allows to print a simple list of supported transliterations
along with some meta information.

=head1 EXPORTS

No symbols are exported by default.

Use either the routine's name or one of the following I<tags> to import
symbols to your namespace.

=over 4

=item B<all>

Import all routines.

=item B<checks>

Import all routines that allow to check if a given transliteration is
supported: translit_supported() and translit_reverse_supported().

=item B<list>

Import translit_list_supported(). (Convenience tag)

=back

=cut

require Exporter;

our @ISA    = qw/Exporter/;
our @EXPORT = qw//;           # Export nothing by default
our @EXPORT_OK = qw/translit_supported translit_reverse_supported
  translit_list_supported/;

our %EXPORT_TAGS = (
    checks => [qw/translit_supported translit_reverse_supported/],
    list   => [qw/translit_list_supported/],
    all    => [@EXPORT_OK]
);

# For convenience, the tables are initialized at the bottom of this file.
our %tables;

# Used internally to retrieve a reference to a single transliteration table.
sub _get_table_reference {
    my $name = shift();

    return unless $name;

    $name = _get_table_id($name);

    foreach my $table ( keys %tables ) {
        return _handle_perl_unicode_bug( $tables{$table} )
          if $table =~ /^$name$/i;
    }

    return;
}

# Handle the "Unicode Bug" affecting code points in the Latin-1 block.
#
# Have a look at perlunicode (section "The 'Unicode Bug'") for details.
sub _handle_perl_unicode_bug {
    my $tbl = shift();

    foreach my $rule ( @{ $tbl->{rules} } ) {
        utf8::upgrade( $rule->{from} );
        utf8::upgrade( $rule->{to} );

        if ( defined( $rule->{context} ) ) {
            utf8::upgrade( $rule->{context}->{before} )
              if defined $rule->{context}->{before};
            utf8::upgrade( $rule->{context}->{after} )
              if defined $rule->{context}->{after};
        }
    }

    return $tbl;
}

=head1 ROUTINES

=head2 translit_supported(I<translit_name>)

Returns true (1), iff I<translit_name> is supported. False (0) otherwise.

=cut

sub translit_supported {
    return ( _get_table_reference( _get_table_id( $_[0] ) ) ? 1 : 0 );
}

=head2 translit_reverse_supported(I<translit_name>)

Returns true (1), iff I<translit_name> is supported and allows reverse
transliteration. False (0) otherwise.

=cut

sub translit_reverse_supported {
    my $table = _get_table_reference( _get_table_id( $_[0] ) );

    croak("Failed to retrieve table for $_[0].") unless ($table);

    return ( ( $table->{reverse} =~ /^true$/ ) ? 1 : 0 );
}

=head2 B<translit_list_supported()>

Prints a list of all supported transliterations to STDOUT, providing the
following information:

  * Name
  * Reversibility
  * Description

The same information is provided in this document as well:

=cut

sub translit_list_supported {
    foreach my $table ( sort keys %tables ) {
        my $t = $tables{$table};
        print "$t->{name}, ",
          ( $t->{reverse} eq "false" ? "not " : "" ),
          "reversible, $t->{desc}\n";
    }
}

=head1 SUPPORTED TRANSLITERATIONS

=over 4

=item Cyrillic

I<ALA-LC RUS>, not reversible, ALA-LC:1997, Cyrillic to Latin, Russian

I<ISO 9>, reversible, ISO 9:1995, Cyrillic to Latin

I<DIN 1460 RUS>, reversible, DIN 1460:1982, Cyrillic to Latin, Russian

I<DIN 1460 UKR>, reversible, DIN 1460:1982, Cyrillic to Latin, Ukrainian

I<DIN 1460 BUL>, reversible, DIN 1460:1982, Cyrillic to Latin, Bulgarian

I<Streamlined System BUL>, not reversible, The Streamlined System: 2006,
Cyrillic to Latin, Bulgarian

I<GOST 7.79 RUS>, reversible, GOST 7.79:2000 (table B), Cyrillic to Latin,
Russian

I<GOST 7.79 RUS OLD>, not reversible, GOST 7.79:2000 (table B), Cyrillic to
Latin with support for Old Russian (pre 1918), Russian

I<GOST 7.79 UKR>, reversible, GOST 7.79:2000 (table B), Cyrillic to Latin,
Ukrainian

=item Greek

I<ISO 843>, not reversible, ISO 843:1997, Greek to Latin

I<DIN 31634>, not reversible, DIN 31634:1982, Greek to Latin

I<Greeklish>, not reversible, Greeklish (Phonetic), Greek to Latin

=item Latin

I<Common CES>, not reversible, Czech without diacritics

I<Common DEU>, not reversible, German without umlauts

I<Common POL>, not reversible, Unaccented Polish

I<Common RON>, not reversible, Romanian without diacritics as commonly used

I<Common SLK>, not reversible, Slovak without diacritics

I<Common SLV>, not reversible, Slovenian without diacritics

=item Arabic

I<Common ARA>, not reversible, Common Romanization of Arabic

=back

=head1 ADDING NEW TRANSLITERATIONS

In case you want to add your own transliteration tables to
L<Lingua::Translit>, have a look at the developer manual included in the
distribution.
An online version is available at
L<http://www.lingua-systems.com/translit/downloads/>.

A template of a transliteration table is provided as well
(F<xml/template.xml>) so you can easily start developing.


=head1 BUGS

None known.

Please report bugs to perl@lingua-systems.com.

=head1 SEE ALSO

L<Lingua::Translit>

L<http://www.lingua-systems.com/translit/>


=head1 CREDITS

Thanks to Dr. Daniel Eiwen, Romanisches Seminar, Universitaet Koeln for his
help on Romanian transliteration.

Thanks to Dmitry Smal and Rusar Publishing for contributing the "ALA-LC RUS"
transliteration table.

Thanks to Ahmed Elsheshtawy for his help implementing the "Common ARA" Arabic
transliteration.

=head1 AUTHORS

Alex Linke <alinke@lingua-systems.com>

Rona Linke <rlinke@lingua-systems.com>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2007-2008 Alex Linke and Rona Linke

Copyright (C) 2009-2014 Lingua-Systems Software GmbH

This module is free software. It may be used, redistributed
and/or modified under the terms of either the GPL v2 or the
Artistic license.

=cut

# Get a table's identifier (based on the table's name)
#   i.e "Common DEU" -> "common_deu"
sub _get_table_id {
    my $name = shift();

    return "" unless $name;

    $name =~ s/\s/_/g;

    return lc($name);
}

# For convenience, the next line is automatically substituted with the set
# of transliteration tables at build time.
%tables;    # PLACEHOLDER

1;

# vim: sts=4 sw=4 ts=4 ai et


lib/Lingua/Translit.pm
package Lingua::Translit;

#
# Copyright (C) 2007-2008 ...
#   Alex Linke <alinke@lingua-systems.com>
#   Rona Linke <rlinke@lingua-systems.com>
# Copyright (C) 2009-2014 Lingua-Systems Software GmbH
#

use strict;
use warnings;

require 5.008;

use Carp qw/croak/;
use Encode qw/encode decode/;

use Lingua::Translit::Tables;

our $VERSION = '0.20';

=pod

=head1 NAME

Lingua::Translit - transliterates text between writing systems

=head1 SYNOPSIS

  use Lingua::Translit;

  my $tr = new Lingua::Translit("ISO 843");

  my $text_tr = $tr->translit("character oriented string");

  if ($tr->can_reverse()) {
    $text_tr = $tr->translit_reverse("character oriented string");
  }

=head1 DESCRIPTION

Lingua::Translit can be used to convert text from one writing system to
another, based on national or international transliteration tables.
Where possible a reverse transliteration is supported.

The term C<transliteration> describes the conversion of text from one
writing system or alphabet to another one.
The conversion is ideally unique, mapping one character to exactly one
character, so the original spelling can be reconstructed.
Practically this is not always the case and one single letter of the
original alpabet can be transcribed as two, three or even more letters.

Furthermore there is more than one transliteration scheme for one writing
system.
Therefore it is an important and necessary information, which scheme will be
or has been used to transliterate a text, to work integrative and be able to
reconstruct the original data.

Reconstruction is a problem though for non-unique transliterations, if no
language specific knowledge is available as the resulting clusters of
letters may be ambigous.
For example, the Greek character "PSI" maps to "ps", but "ps" could also
result from the sequence "PI", "SIGMA" since "PI" maps to "p" and "SIGMA"
maps to s.
If a transliteration table leads to ambigous conversions, the provided
table cannot be used reverse.

Otherwise the table can be used in both directions, if appreciated.
So if ISO 9 is originally created to convert Cyrillic letters to
the Latin alphabet, the reverse transliteration will transform Latin
letters to Cyrillic.

=head1 METHODS

=head2 new(I<"name of table">)

Initializes an object with the specific transliteration table, e.g. "ISO 9".

=cut

sub new {
    my $class = shift();
    my $name  = shift();

    my $self;

    # Assure that a table name was set
    croak("No transliteration name given.") unless $name;

    # Stay compatible with programs that use Lingua::Translit < 0.05
    if ( $name =~ /^DIN 5008$/i ) {
        $name = "Common DEU";
    }

    my $table = Lingua::Translit::Tables::_get_table_reference($name);

    # Check that a table reference was assigned to the object
    croak("No table found for $name.") unless $table;

    # Assure the table's data is complete
    croak("$name table: missing 'name'")    unless defined $table->{name};
    croak("$name table: missing 'desc'")    unless defined $table->{desc};
    croak("$name table: missing 'reverse'") unless defined $table->{reverse};
    croak("$name table: missing 'rules'")   unless defined $table->{rules};

    # Copy over the table's data
    $self->{name}  = $table->{name};
    $self->{desc}  = $table->{desc};
    $self->{rules} = $table->{rules};

    # Set a truth value of the transliteration's reversibility according to
    # the natural language string in the original transliteration table
    $self->{reverse} = ( $table->{reverse} =~ /^true$/i ) ? 1 : 0;

    undef($table);

    return bless $self, $class;
}

=head2 translit(I<"character oriented string">)

Transliterates the given text according to the object's transliteration
table.
Returns the transliterated text.

=cut

sub translit {
    my $self = shift();
    my $text = shift();

    # Return if no input was given
    return unless $text;

    my $utf8_flag_on = Encode::is_utf8($text);

    unless ($utf8_flag_on) {
        $text = decode( "UTF-8", $text );
    }

    foreach my $rule ( @{ $self->{rules} } ) {
        if ( defined $rule->{context} ) {
            my $c = $rule->{context};

            # single context rules
            if ( defined $c->{before} && !defined $c->{after} ) {
                $text =~ s/$rule->{from}(?=$c->{before})/$rule->{to}/g;
            }
            elsif ( defined $c->{after} && !defined $c->{before} ) {
                $text =~ s/(?<=$c->{after})$rule->{from}/$rule->{to}/g;
            }

            # double context rules: logical "inbetween"
            elsif ( defined $c->{before} && defined $c->{after} ) {
                $text =~ s/
                (?<=$c->{after})$rule->{from}(?=$c->{before})
                /$rule->{to}/gx;
            }

            else {
                croak("incomplete rule context");
            }
        }
        else {
            $text =~ s/$rule->{from}/$rule->{to}/g;
        }
    }

    unless ($utf8_flag_on) {
        return encode( "UTF-8", $text );
    }
    else {
        return $text;
    }
}

=head2 translit_reverse(I<"character oriented string">)

Transliterates the given text according to the object's transliteration
table, but uses it the other way round. For example table ISO 9 is a
transliteration scheme for the converion of Cyrillic letters to the Latin
alphabet. So if used reverse, Latin letters will be mapped to Cyrillic ones.

Returns the transliterated text.

=cut

sub translit_reverse {
    my $self = shift();
    my $text = shift();

    # Return if no input was given
    return unless $text;

    # Is this transliteration reversible?
    croak("$self->{name} cannot be reversed") unless $self->{reverse};

    my $utf8_flag_on = Encode::is_utf8($text);

    unless ($utf8_flag_on) {
        $text = decode( "UTF-8", $text );
    }

    foreach my $rule ( @{ $self->{rules} } ) {
        if ( defined $rule->{context} ) {
            my $c = $rule->{context};

            # single context rules
            if ( defined $c->{before} && !defined $c->{after} ) {
                $text =~ s/$rule->{to}(?=$c->{before})/$rule->{from}/g;
            }
            elsif ( defined $c->{after} && !defined $c->{before} ) {
                $text =~ s/(?<=$c->{after})$rule->{to}/$rule->{from}/g;
            }

            # double context rules: logical "inbetween"
            elsif ( defined $c->{before} && defined $c->{after} ) {
                $text =~ s/
                    (?<=$c->{after})$rule->{to}(?=$c->{before})
                    /$rule->{from}/gx;
            }

            else {
                croak("incomplete rule context");
            }
        }
        else {
            $text =~ s/$rule->{to}/$rule->{from}/g;
        }
    }

    unless ($utf8_flag_on) {
        return encode( "UTF-8", $text );
    }
    else {
        return $text;
    }
}

=head2 can_reverse()

Returns true (1), iff reverse transliteration is possible.
False (0) otherwise.

=cut

sub can_reverse {
    return $_[0]->{reverse};
}

=head2 name()

Returns the name of the chosen transliteration table, e.g. "ISO 9".

=cut

sub name {
    return $_[0]->{name};
}

=head2 desc()

Returns a description for the transliteration,
e.g. "ISO 9:1995, Cyrillic to Latin".

=cut

sub desc {
    return $_[0]->{desc};
}

=head1 SUPPORTED TRANSLITERATIONS

=over 4

=item Cyrillic

I<ALA-LC RUS>, not reversible, ALA-LC:1997, Cyrillic to Latin, Russian

I<ISO 9>, reversible, ISO 9:1995, Cyrillic to Latin

I<DIN 1460 RUS>, reversible, DIN 1460:1982, Cyrillic to Latin, Russian

I<DIN 1460 UKR>, reversible, DIN 1460:1982, Cyrillic to Latin, Ukrainian

I<DIN 1460 BUL>, reversible, DIN 1460:1982, Cyrillic to Latin, Bulgarian

I<Streamlined System BUL>, not reversible, The Streamlined System: 2006,
Cyrillic to Latin, Bulgarian

I<GOST 7.79 RUS>, reversible, GOST 7.79:2000 (table B), Cyrillic to Latin,
Russian

I<GOST 7.79 RUS OLD>, not reversible, GOST 7.79:2000 (table B), Cyrillic to
Latin with support for Old Russian (pre 1918), Russian

I<GOST 7.79 UKR>, reversible, GOST 7.79:2000 (table B), Cyrillic to Latin,
Ukrainian

=item Greek

I<ISO 843>, not reversible, ISO 843:1997, Greek to Latin

I<DIN 31634>, not reversible, DIN 31634:1982, Greek to Latin

I<Greeklish>, not reversible, Greeklish (Phonetic), Greek to Latin

=item Latin

I<Common CES>, not reversible, Czech without diacritics

I<Common DEU>, not reversible, German without umlauts

I<Common POL>, not reversible, Unaccented Polish

I<Common RON>, not reversible, Romanian without diacritics as commonly used

I<Common SLK>, not reversible, Slovak without diacritics

I<Common SLV>, not reversible, Slovenian without diacritics

=back

=head1 ADDING NEW TRANSLITERATIONS

In case you want to add your own transliteration tables to
L<Lingua::Translit>, have a look at the developer manual included in the
distribution.
An online version is available at
L<http://www.lingua-systems.com/translit/downloads/>.

A template of a transliteration table is provided as well
(F<xml/template.xml>) so you can easily start developing.


=head1 RESTRICTIONS

L<Lingua::Translit> is suited to handle B<Unicode> and utilizes comparisons
and regular expressions that rely on B<code points>.
Therefore, any input is supposed to be B<character oriented>
(C<use utf8;>, ...) instead of byte oriented.

However, if your data is byte oriented, be sure to pass it
B<UTF-8 encoded> to translit() and/or translit_reverse() - it will be
converted internally.

=head1 BUGS

None known.

Please report bugs to perl@lingua-systems.com.

=head1 SEE ALSO

L<Lingua::Translit::Tables>, L<Encode>, L<perlunicode>

L<translit(1)>

L<http://www.lingua-systems.com/translit/>

=head1 CREDITS

Thanks to Dr. Daniel Eiwen, Romanisches Seminar, Universitaet Koeln for his
help on Romanian transliteration.

Thanks to Dmitry Smal and Rusar Publishing for contributing the "ALA-LC RUS"
transliteration table.

=head1 AUTHORS

Alex Linke <alinke@lingua-systems.com>

Rona Linke <rlinke@lingua-systems.com>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2007-2008 Alex Linke and Rona Linke

Copyright (C) 2009-2014 Lingua-Systems Software GmbH

This module is free software. It may be used, redistributed
and/or modified under the terms of either the GPL v2 or the
Artistic license.

=cut

1;

# vim: sts=4 sw=4 ts=4 ai et
